python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,157 @@
+                      'possibly-undefined',
+                      'Reported when a statement uses a variable that might not be assigned.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    possibly_undefined = set()
+
+    @check_messages('possibly-undefined')
+    def visit_name(self, node):
+        if self._check_possibly_undefined(node):
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _check_possibly_undefined(self, node: astroid.Name) -> bool:
+        """Returns True if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        return node in self.possibly_undefined
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `possibly_undefined` if it might not be defined.
+        """
+        facts = {}
+        blocks = self._get_blocks_po(node)
+
+        all_assigns = self._get_assigns(blocks)
+        for block in blocks:
+            facts[block] = {}
+            facts[block]['out'] = all_assigns.copy()
+
+        worklist = blocks[:]
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [facts[p.source]['out'] for p in b.predecessors if p.source in facts]
+            if outs == []:
+                facts[b]['in'] = set()
+            else:
+                facts[b]['in'] = set.intersection(*outs)
+            temp = self._transfer(b, facts[b]['in'], all_assigns)
+            if temp != facts[b]['out']:
+                facts[b]['out'] = temp
+                successors = set([succ.target for succ in b.successors])
+                worklist = list(set(worklist).union(successors))
+
+    def _transfer(self, block: CFGBlock, in_facts: Set[str], all_assigns: Set[str]) -> Set[str]:
+        gen = set()
+        kill = set()
+        for statement in block.statements:
+            if isinstance(statement, astroid.Assign):
+                gen.update(self._get_targets(statement))
+            elif isinstance(statement, astroid.AnnAssign) and hasattr(statement.target, 'name'):
+                gen.add(statement.target)
+            elif isinstance(statement, astroid.Delete):
+                kill.update(self._get_targets(statement))
+            self._analyze_statement(statement, gen.union(in_facts.difference(kill)), all_assigns)
+        return gen.union(in_facts.difference(kill))
+
+    def _analyze_statement(self, node: NodeNG, facts: Set[str], local_vars: Set[str]) -> None:
+        if isinstance(node, astroid.Name):
+            name = node.name
+            if name in local_vars and name not in facts and not self._is_function_name(node):
```

Comment:
There shouldn't need to be a special case for names of functions. It seems like `_transfer` above isn't handling the possibility that a statement in a block is a `FunctionDef`, which is another way of introducing a local name.

Reply:
The `_transfer` function loops over all the statements in the block, so it is possible that you come across a `FunctionDef` node (nested function). However, The reason behind having this extra check is that a `Name` node could be the name of a function in a function call. For example the ast representation of  `print(x)` would be `Call(n=Name('print'),args=Name('x'))`. This means that the `name` variable in the code above could be storing `'print'`, a function name. It gets tricky when the function name is equivalent to a local variable name. In that case, there would be no difference (in the `Name` node) between a function name and a variable name.